pytket/extensions/cutensornet/tnstate/ttn.py

```diff
@@ -0,0 +1,854 @@
+
+            Q, R = tensor.decompose(
+                node_bonds + "->" + Q_bonds + "," + R_bonds,
+                leaf_node.tensor,
+                method=tensor.QRMethod(),
+                options=options,
+            )
+            # Note: Since R is not contracted with any other tensor, we cannot update
+            #   the leaf node to Q. That'd change the state represented by the TTN.
+
+        # Otherwise, if ``unsafe`` is enabled, update the last tensor to Q
+        elif unsafe:
+            self.nodes[target_path[:-1]].tensor = Q
+            self.nodes[target_path[:-1]].canonical_form = target_path[-1]
+
+            self._logger.debug(f"Node canonicalised (unsafe!). Shape: {Q.shape}")
+
+        self._logger.debug(
+            f"Finished canonicalisation. Returning R tensor of shape {R.shape}"
+        )
+        return R
+
+    def vdot(self, other: TTN) -> complex:  # type: ignore
+        """Obtain the inner product of the two TTN: ``<self|other>``.
+
+        It can be used to compute the squared norm of a TTN ``ttn`` as
+        ``ttn.vdot(ttn)``. The tensors within the TTN are not modified.
+
+        Note:
+            The state that is conjugated is ``self``.
+
+        Args:
+            other: The other TTN.
+
+        Returns:
+            The resulting complex number.
+
+        Raises:
+            RuntimeError: If the two TTNs do not have the same qubits.
+            RuntimeError: If the ``CuTensorNetHandle`` is out of scope.
+        """
+        if self._lib._is_destroyed:
+            raise RuntimeError(
+                "The cuTensorNet library handle is out of scope.",
+                "See the documentation of update_libhandle and CuTensorNetHandle.",
+            )
+
+        if len(self.qubit_position) != len(other.qubit_position):
+            raise RuntimeError("Number of qubits do not match.")
+        if self.get_qubits() != other.get_qubits():
+            raise RuntimeError(
+                "The sets of qubits are not the same."
+                "\n\tself has {self.get_qubits()}"
+                "\n\tother has {other.get_qubits()}"
+            )
+        if len(self.qubit_position) == 0:
+            raise RuntimeError("There are no qubits in the TTN.")
+
+        self._logger.debug("Applying vdot between two TTNs.")
+
+        # We convert both TTNs to their interleaved representation and
+        # contract them using cuQuantum. A single sample is enough for
+        # contraction path optimisation, since there is little to optimise.
+        ttn1 = self.get_interleaved_representation(conj=True)
+        ttn2 = other.get_interleaved_representation(conj=False)
+        interleaved_rep = ttn1 + ttn2 + [[]]  # Discards dim=1 bonds with []
+        result = cq.contract(
+            *interleaved_rep,
+            options={"handle": self._lib.handle, "device_id": self._lib.device_id},
+            optimize={"samples": 1},  # There is little to no optimisation to be done

```

Comment:
We have to run at least one, haven't we? Would be good to skip altogether, taken it has to be done on CPU.

Reply:
Actually, I had assumed that you needed at least one sample, but no, if you set it to `0` it still works. This are the samples of the hyperoptimizer, but as far as I can tell the path returned by the first pass (with no hyperoptimization) is already reasonable (I have inspected the contraction path of a couple of examples), so I'll set this to `"samples": 0`.

In the case of TTN it's a bit more involved to provide a contraction path explicitly. The main reason being that I intend to allow `vdot` of TTNs where the arrangement of qubits in leaves may be different (this is not allowed in MPS). This is immediately supported if we let `cq.contract` figure out the path, but an efficient contraction path cannot be hardcoded anymore; I'd need to use some simple greedy algorithm, and I expect cq.contract will do better and faster on that front.

Reply:
I see, makes sense!

Reply:
Done